clang/test/Driver/fveclib.c

```diff
@@ -36,16 +36,23 @@
 /* Verify that the correct vector library is passed to LTO flags. */
 
 // RUN: %clang -### --target=x86_64-unknown-linux-gnu -fveclib=LIBMVEC -flto %s 2>&1 | FileCheck --check-prefix=CHECK-LTO-LIBMVEC %s
+// CHECK-LTO-LIBMVEC: "-fmath-errno"
 // CHECK-LTO-LIBMVEC: "-plugin-opt=-vector-library=LIBMVEC-X86"
 
 // RUN: %clang -### --target=powerpc64-unknown-linux-gnu -fveclib=MASSV -flto %s 2>&1 | FileCheck --check-prefix=CHECK-LTO-MASSV %s
+// CHECK-LTO-MASSV: "-fmath-errno"
 // CHECK-LTO-MASSV: "-plugin-opt=-vector-library=MASSV"
 
 // RUN: %clang -### --target=x86_64-unknown-linux-gnu -fveclib=SVML -flto %s 2>&1 | FileCheck --check-prefix=CHECK-LTO-SVML %s
+// CHECK-LTO-SVML: "-fmath-errno"
 // CHECK-LTO-SVML: "-plugin-opt=-vector-library=SVML"
 
 // RUN: %clang -### --target=aarch64-linux-gnu -fveclib=SLEEF -flto %s 2>&1 | FileCheck --check-prefix=CHECK-LTO-SLEEF %s
+// CHECK-LTO-SLEEF-NOT: "-fmath-errno"
 // CHECK-LTO-SLEEF: "-plugin-opt=-vector-library=sleefgnuabi"
+// CHECK-LTO-SLEEF-NOT: "-fmath-errno"
 
 // RUN: %clang -### --target=aarch64-linux-gnu -fveclib=ArmPL -flto %s 2>&1 | FileCheck --check-prefix=CHECK-LTO-ARMPL %s
+// CHECK-LTO-ARMPL-NOT: "-fmath-errno"
```

Comment:
I think it might be worth to add tests for "-fveclib=ArmPL -fmath-errno" and similar for sleefm to ensure that the order of flags in properly honoured.

Reply:
I agree but I think it should be stronger and we should emit a warning diagnostic that says the potential to use the specified vector library will be hampered because -fmath-errnor has been requested.

I'm suggesting a warning rather than remark because we're basically saying the user has likely made a mistake because it doesn't make sense to request both features.

Reply:
Added the extra tests for `-fmath-errno` after `-fveclib`. 

If we want to add a warning there's a few cases to consider where errno could be re-enabled: 

- `-fveclib=ArmPL -fmath-errno` (the obvious case)
- `-fveclib=ArmPL -fno-fast-math`
- `-fveclib=ArmPL -ffp-model=strict`

Would a warning be expected for all of these?

Reply:
I think it'll also need to track which option/arg caused `-fmath-errno` to be set (otherwise the warning would be somewhat unhelpful in some cases).

Reply:
I'm not too worried about that, it's not really the responsibility of this warning to explain why math-errno is in effect.

Reply:
Ah well, I did add it to the diagnostic :sweat_smile:  (I don't think it adds much complexity, and may be helpful), but I don't mind removing it :slightly_smiling_face:

Reply:
Thanks @MacDue for these changes, personally I find the warning in the current form helpful, although it  feels to me like it combines 2 things:

1. warning that math-errno got re-enabled
2. loop vectorizer optimization remark suggesting to use -fno-math-errno to enable vectorization loops calling math functions.

Do we have the pt 2 already?

Reply:
I've followed @paulwalker-arm's suggestion and removed the vectorizer specific phrasing from the warning, so it's now:
> math errno enabled by '-ffp-model=strict' after it was implicitly disabled by '-fveclib=ArmPL', this may limit the utilization of the vector library [-Wmath-errno-enabled-with-veclib]

Which I think addresses the "combines 2 things" issue. I'm not sure if pt 2 exists. Is `errno` a concept that exists in the loop vectorizer, other than as a memory effect? It seems more C/C++ language specific.

That said, I did spot: https://github.com/llvm/llvm-project/blob/e13f1d1daf9b76134c3585e8250941920bdf3da6/llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp#L908-L918